docs: update contributing document with information about our workflows (#2176)
influxdata/flux: CONTRIBUTING.md

# Contributing to Flux

## Community Slack Channel
When contributing to Flux, it is a good idea to join the [community Slack channel](https://www.influxdata.com/slack).
The development team for Flux is in the `#flux` channel and we will be able to answer any questions or give any recommendations for development work from there.

## Filing issues
Filing issues on GitHub is one of the easiest and most useful ways to contribute to Flux.

> We value every request and we intend to triage every community issue that is created within a week after it has been created.

Reviewer: ```suggestion
We value every request and we intend to triage every community issue within a week after it has been created.
```

If it takes us longer than a week, then please try to contact us in the community Slack channel.

Before you file an issue, please search existing issues in case it has already been filed or perhaps even resolved.

### Bug reports
When submitting a bug report, please include the following:

- Full details of your operating system (or distribution) e.g. 64-bit Ubuntu 14.04.
- The version of Flux you are running.
- Whether you installed it using a pre-built package or built it from source.

> - A small test case, if applicable, that demonstrates the issues.

Reviewer: ```suggestion
- A small test case that demonstrates the issue or steps to reproduce the issue
```
Telling us how to reproduce the issue could be just as helpful here.

> Remember the golden rule of bug reports: **The easier you make it for us to reproduce the problem, the faster it will get fixed.

Reviewer: Yeah, exactly.

> be closed if we don't understand what you would like to see added to Flux.

Reviewer: Would we not first ask for clarification in this case?

## Contributing to the source code

### Finding an issue
The Flux team regularly adds the [community](https://github.com/influxdata/flux/issues?q=is%3Aopen+is%3Aissue+label%3Acommunity) label to issues that we think would be accessible for a community member to take.

> Before starting to work on an issue, please inform us on GitHub about your intention to work on the issue.

Reviewer: ```suggestion
Before starting to work on an issue, please inform us on GitHub about your intention to work on the issue by leaving a comment on the issue and assigning it to yourself.
```

Author: I actually don't think they can assign it to themselves. I'll have to fix this.

For rationale, check out the post [The Case Against Third Party Libraries](http://blog.gopheracademy.com/advent-2014/case-against-3pl/).

### Writing unit tests
Unit tests for Go code should be in the same package as the code itself.
It should be in a file named `filename_test.go` with the package name `mypackage_test` as the name of the package.
The tests should test functionality through public functions and methods.

If access to internal members or functions is needed, then the test file should be named `filename_internal_test.go` and use the same package name as the package itself.
The tests written in this file will have access to the internals of the package.
This file can also be used to create testing-specific functions that can be exposed to the tests in `mypackage_test`.

> Pull requests require tests to be written for them and must pass our code coverage checks.

Reviewer: This might be a good place to link to a longer article about writing end-to-end tests that includes a code sample, once we have that.

### Public functions and methods
When writing code, try to avoid making functions and methods used by the code public.
When creating a library that needs to be used by multiple packages within the Flux repository, an internal package is preferred.
A public function or method should only be used if using the Flux library would require the function or method to be public.

### Comments and documentation
Every new public function or method requires a valid GoDoc.
Comments within the code should be written with proper capitalization and punctuation at the end of the sentence.

> 

Reviewer: Can we link to Go's documentation on how to write proper comments? I think this might be a good one: https://blog.golang.org/godoc-documenting-go-code

Author: Yes and I should find a place to link effective go somewhere.

### Updating markdown files for documentation
The Flux team uses markdown for our documentation.

> When writing markdown, please use the following guidelines.

Reviewer: ```suggestion
When writing markdown, please use the following guidelines:
```

- Each sentence in a paragraph is on its own line.
- Use the `#` symbol for header sections.
- Use the `-` symbol for bullet points.

## Submitting a pull request
To submit a pull request you should fork the Flux repository and make your change on a feature branch of your fork.
Then generate a pull request from your branch against **master** of the Flux repository.

> Include in your pull request details of your change -- the **why** *and* the **how** -- as well as the testing you performed.

Reviewer: @jsternberg, you write really good PR details. Maybe link here to a recent PR that you opened with a detailed comment for reference.

Author: It's my opinion that we shouldn't refer to a specific PR for this. I think it might be good to give some kind of guidelines, but I can't think of anything that would work well in a readme.

Reviewer: Guidelines are a good idea. We can keep it how it is for now and then add these later when we see what contributors come up with.

There will usually be some back and forth as we finalize the change, but once that completes, it may be merged.
When making changes based on the code review, please add new commits and please refrain from amending the original commit.
This helps us keep track of the changes that have been made during the course of the review.

When your change is ready to be merged, the Flux developer who is reviewing your code will approve the pull request and squash your commits into a single commit.
The first commit in your changeset will be used as the final commit message.
This first commit must follow the _conventional commit message_ format: https://www.conventionalcommits.org/en/v1.0.0-beta.3/.
Please use this commit message format for commits that will be visible in influxdata/flux history.

If your pull request only requires trivial changes before being approved and merged, the reviewer may make the change themselves and push to the branch in your fork.
This makes it easier for us to fix a small nitpick such as formatting without requiring an additional back and forth.
If the changes are non-trivial, we will always ask before changing any code.

> If you do not want us to do this at all under any circumstance, then you can disable this by unselecting **Allow edits from maintainers** in the pull request itself.

Reviewer: That's a really good addition. 👍

To assist in review for the PR, please add the following to your pull request comment:

```md
- [ ] Sign [CLA](https://www.influxdata.com/legal/cla/) (if not already signed)
```

### Signing the CLA
In order to contribute back to Flux, you must sign the
[InfluxData Contributor License Agreement](https://www.influxdata.com/legal/cla/) (CLA).

> 

Reviewer: This section should probably go before the PR section since they have to do this before opening the PR.
